### ISSUE TITLE
[Bug]: OpenSearch Vector Store have a lot of connection problem after #11513

### ISSUE THREAD

**Issue by rendyfebry** [run-llama/llama_index#13692]
### Bug Description

I noticed that after #11513, my OpenSearch vector store connection became unstable, and we can confirm this by trending bug reports related to this issue recently. (#13358, #11661, etc)

- This event loop is already running.
- ConnectionError(Timeout context manager should be used inside a task) caused by: RuntimeError(Timeout context manager should be used inside a task)
- etc

I also see it suggested to use asyncio, but even with that we can still get the errors
```
import nest_asyncio
nest_asyncio.apply()
```

I also notice, this error getting worse when we use LlamaIndex with FastAPI,

### Version

latest

### Steps to Reproduce

Just use OpenSearch vector store long enough that you'll notice it's become unstable.

**Comment by logan-markewich** [comment]
It would be nice to support async in this vector store. If you have any ideas for improvements, open to PRs 😇

**Comment by rendyfebry** [comment]
@logan-markewich I Agree with the async, but I think would be great if the changes were backward compatible or at least made them optional, so it won't break an existing implementation.

Unfortunately, I'm not really good with OpenSearch connection or python async handling, so I can't help with this one.

**Comment by der-ofenmeister** [comment]
I have written a patched Opensearch class to handle async and connection issues do to this - happy to send a PR, although I'm not sure how compatible it will be (should probably be good for llama-index v0.10+)

**Comment by GeorgeDittmar** [comment]
We are running into similar issues. I had to basically rewrite the Opensearch support to go back to using the sync version of opensearch-py to get it to run in our fast api service. I could put in a PR for allowing a flag to use async or not possibly?

**Comment by ulan-yisaev** [comment]
I also encountered similar connection issues when using OpenSearch with FastAPI. I think the patch by @der-ofenmeister or the flag to allow async/sync choice suggested by @GeorgeDittmar would be a great solution. It would be much better to avoid using workarounds like nest_asyncio.

Looking forward to the patch!

**Comment by GeorgeDittmar** [comment]
yeah nest_asyncio didnt even work for us on our service but even then I wouldnt trust it for a prod solution. I'll maybe start work on a flag approach unless @der-ofenmeister puts their patch in first

**Comment by ulan-yisaev** [comment]
@GeorgeDittmar , I’ve experienced similar connection problems when using OpenSearch with FastAPI as well. I agree with avoiding nest_asyncio for production use; it didn't work well in our setup either.

**Comment by ulan-yisaev** [comment]
Hi @der-ofenmeister, would you be able to share the details of the patch you mentioned? I'd love to test it out, as we're facing similar issues and looking for a more stable solution than nest_asyncio.

Thanks in advance!

**Comment by ulan-yisaev** [comment]
It wasn't as straightforward as expected :). I had to implement custom versions of OpensearchVectorStore and OpensearchVectorClient to switch the async operations to synchronous ones. After doing that, it seems to be working now.

**Comment by GeorgeDittmar** [comment]
Yeah thats exactly what I had to do on my end sadly.

**Comment by logan-markewich** [comment]
Seems like no one here is willng to submit a PR for this (either that, or its already resolved)

Going to close this out. If others are having issues. please open a PR
